tiledb/sm/filesystem/temporary_directory.h

```diff
@@ -0,0 +1,56 @@
+/**
+ * @file   temporary_directory.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements static function filesystem::make_temporary_directory.
+ */
+
+#ifndef TILEDB_TEMPORARY_DIRECTORY_H
+#define TILEDB_TEMPORARY_DIRECTORY_H
+
+#include <filesystem>
+#include <mutex>
+#include "tiledb/common/random/random.h"
+
+namespace tiledb::sm {
+
+static std::mutex mtx_;
+
+static const std::string& make_temporary_directory(
```

Comment:
I don't see that this class actually does a file operation to make the directory. If creating the directory fails, it should throw.

This function should probably be renamed `temporary_directory_name` and called by a function of the current name.